non-strict export with dynamic shapes

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115387

Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

avikchaudhuri committed:
non-strict export with dynamic shapes



Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

[ghstack-poisoned]

zhxchen17 commented on torch/_export/__init__.py:
```diff
@@ -792,15 +792,21 @@ def strip_root(x):
-            {},
-            [],
+            range_constraints,
+            equality_constraints,
```
not blocking but I feel maybe we could stop returning equality constraints as they will be removed and no one actively use them?

avikchaudhuri replied:
I want to preserve behavior for now to minimize blast radius. I can zero out all equality constraints in a different PR (I was under the impression this was already done...)

zhxchen17 reviewed (commented):
Looks good to me overall, will let @tugsbayasgalan comment on dynamic shape part

avikchaudhuri committed:
Update on "non-strict export with dynamic shapes"


Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -0,0 +1,145 @@
+            tensor_constraints[constraint.shared.t_id][constraint.shared.dim] = constraint
+
+    code = nn_module.forward.__code__
+    co_fields = {
```
Just curious: why do we need these in the shape_env

avikchaudhuri replied:
I wanted the shape env to be as much the same as what's produced by strict, and I noticed that this is populated in strict mode.

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -0,0 +1,145 @@
+        params = inspect.signature(nn_module.forward).parameters
+        dim_name_to_sources: Dict[str, List[Source]] = defaultdict(list)
+        fake_args = tuple(
+            fake_tree(
```
nit: Why can't we use pytree here?

avikchaudhuri replied:
Yeah that would be nice, but unfortunately cannot. We're building sources too as we descend. (I can check if there's a pytree util that can do additional bookkeeping as we descend...)

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -0,0 +1,145 @@
+        ignore_static=False,
+    )
+    shape_env.frozen = True
+    shape_env.dim_constraints.solve()
```
Should we also move the constraint discovery too?

avikchaudhuri replied:
Yes I will, separately. It should follow the same pattern as here, but don't want to deal with potential strict test failures in this PR.

avikchaudhuri replied:
Wait, I'm not sure what you mean...I was assuming you mean, can we move around steps in strict mode to more closely reflect what's going on here.

As for constraint discovery, well that is essentially `.solve()` + pretty-printing the suggestions. And yes, I forgot to add that. There's actually some non-trivial logic around constraint violation errors that we need to replicate from `eval_frame`, including turning them into user errors, catching them, adding some message related to constraint discovery, etc.

tugsbayasgalan replied:
Yep I was referring to .solve() + pretty printing the suggestions.

tugsbayasgalan reviewed (commented):
Awesome :) Just have few questions

avikchaudhuri committed:
Update on "non-strict export with dynamic shapes"


Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "non-strict export with dynamic shapes"


Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "non-strict export with dynamic shapes"


Differential Revision: [D51959229](https://our.internmc.facebook.com/intern/diff/D51959229/)

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge

avikchaudhuri commented:
@pytorchbot merge

zhxchen17 commented:
@pytorchbot rebase
